### PR TITLE
Replace CountedPtr by std::shared_ptr (2nd part)

### PR DESCRIPTION
**Pull request by gervandiepen**
Ths PR replaces CountedPtr by std::shared_ptr in the casa and tables module, except the IO part which was done in PR #1284.

### CODE REVIEW

File: casa/Logging/LogSink.h [hunk 1]
```diff
@@ -302,15 +300,15 @@ class LogSink : public LogSinkInterface
...
+  std::shared_ptr<LogSinkInterface> local_sink_p;
+  static std::shared_ptr<LsiIntermediate> * global_sink_p;
   static std::once_flag theirCallOnceFlag;
 
   // The following is a reference to the global sink. It is created to
   // ensure that the global sink is not destroyed before the last local
   // reference to it is destroyed. This can happen if you have a static
   // LogSink (or LogIO).
-  CountedPtr<LsiIntermediate> local_ref_to_global_p;
+  std::shared_ptr<LsiIntermediate> local_ref_to_global_p;
```
Comment (mnijhuis-tos) [inline casa/Logging/LogSink.h:311]: I don't understand how "This can happen if you have a static LogSink (or LogIO)": The static member should be destroyed only at the end of the application, after all regular objects are already destroyed.

File: casa/Utilities/ValType.cc [hunk 2]
```diff
@@ -460,39 +460,39 @@ ObjCompareFunc* ValType::getCmpFunc (DataType dt)
 }
 
 //# Get the comparison object.
-CountedPtr<BaseCompare> ValType::getCmpObj (DataType dt)
+std::shared_ptr<BaseCompare> ValType::getCmpObj (DataType dt)
 {
     switch (dt) {
     case TpBool:
-        return new ObjCompare<Bool>();
+        return std::shared_ptr<BaseCompare> (new ObjCompare<Bool>());
```
Comment (mnijhuis-tos) [inline casa/Utilities/ValType.cc:467]: Optional: Use `std::make_shared` in all `case`s.

File: tables/TaQL/ExprAggrNodeArray.cc [hunk 3]
```diff
@@ -68,25 +68,25 @@ namespace casacore { //# NAMESPACE CASACORE - BEGIN
     return itsFunc->isLazy();
   }
 
-  CountedPtr<TableExprGroupFuncBase> TableExprAggrNodeArray::doMakeGroupAggrFunc()
+  std::shared_ptr<TableExprGroupFuncBase> TableExprAggrNodeArray::doMakeGroupAggrFunc()
   {
     if (funcType() == TableExprFuncNode::gexpridFUNC) {
-      return new TableExprGroupExprId(this);
+      return std::shared_ptr<TableExprGroupFuncBase> (new TableExprGroupExprId(this));
```
Comment (mnijhuis-tos) [inline tables/TaQL/ExprAggrNodeArray.cc:74]: Optional: Use `return std::make_shared<...>(...);` in this function.

File: tables/TaQL/ExprNode.cc [hunk 4]
```diff
@@ -1494,20 +1494,17 @@ void TableExprNode::adaptUnit (const Unit& unit)
 TableExprNode TableExprNode::newRownrNode (const TableExprInfo& tableInfo,
                                            uInt origin)
 {
-    TENShPtr tsnptr = new TableExprNodeRownr (tableInfo, origin);
-    return tsnptr;
+    return TENShPtr(new TableExprNodeRownr (tableInfo, origin));
```
Comment (mnijhuis-tos) [inline tables/TaQL/ExprNode.cc:1497]: Optional:
```suggestion
    return std::make_shared<TableExprNodeRownr>(tableInfo, origin);
```
Similar improvements apply to the return statements in the functions below.

File: tables/TaQL/ExprNodeArray.cc [hunk 5]
```diff
@@ -66,47 +66,47 @@ TENShPtr TableExprNodeArray::makeConstantScalar()
       {
         MArray<Bool> arr = getArrayBool(0);
         if (arr.size() == 1) {
-          return new TableExprNodeConstBool (arr.array().data()[0]);
+          return TENShPtr(new TableExprNodeConstBool (arr.array().data()[0]));
```
Comment (mnijhuis-tos) [inline tables/TaQL/ExprNodeArray.cc:69]: Optional, also for the `return` statements in the other `case`s.
```suggestion
          return std::make_shared<TableExprNodeConstBool>(arr.array().data()[0]);
```

File: tables/TaQL/ExprNodeRep.h [hunk 6]
```diff
@@ -52,7 +52,7 @@ template<class T> class Block;
 
 //# Define a shared pointer to the Rep class.
 class TableExprNodeRep;
-typedef CountedPtr<TableExprNodeRep> TENShPtr;
+typedef std::shared_ptr<TableExprNodeRep> TENShPtr;
```
Comment (mnijhuis-tos) [inline tables/TaQL/ExprNodeRep.h:55]: Optional: Within team Schaap we are moving away from typedefs for std::shared_ptr's, since they have little value and make the code harder to read. Since you're updating code that uses `TENShPtr` anyway, please consider removing this typedef. When using `std::make_shared`, you don't even need it.

File: tables/TaQL/ExprNodeSet.cc [hunk 7]
```diff
@@ -217,7 +217,7 @@ TENShPtr TableExprNodeSet::setOrArray() const
         for (size_t i=0; i<n; i++) {
               if (! itsElems[i]->unit().empty()) {
                   if (! q.isConform (itsElems[i]->unit())) {
-                      return new TableExprNodeSet (*this);
+                    return TENShPtr(new TableExprNodeSet (*this));
```
Comment (mnijhuis-tos) [inline tables/TaQL/ExprNodeSet.cc:220]: Optional:
```suggestion
                    return std::make_shared<TableExprNodeSet>(*this);
```

File: tables/TaQL/ExprNodeSet.cc [hunk 8]
```diff
@@ -244,14 +244,15 @@ TENShPtr TableExprNodeSet::setOrArray() const
         // Set ndim and shape if those are constant.
     }
     TableExprNodeSet* set = new TableExprNodeSet (*this);
+    TENShPtr pset(set);
```
Comment (mnijhuis-tos) [inline tables/TaQL/ExprNodeSet.cc:247]:
```suggestion
    auto set = std::make_shared<TableExprNodeSet>(*this);
```

File: tables/TaQL/ExprNodeSet.cc [hunk 9]
```diff
@@ -244,14 +244,15 @@ TENShPtr TableExprNodeSet::setOrArray() const
...
     if (itsBounded) {
         // Set the type to VTArray; the getArray functions
         // will convert the set to an array for each row.
         // Set the shape and/or dimensionaly if constant.
         set->setValueType (VTArray);
         set->setShape();
     }
-    return set;
+    return pset;
```
Comment (mnijhuis-tos) [inline tables/TaQL/ExprNodeSet.cc:255]: When updating line 246, you can revert this change.
```suggestion
    return set;
```

File: tables/TaQL/TaQLNodeHandler.cc [hunk 10]
```diff
@@ -473,23 +473,26 @@ namespace casacore { //# NAMESPACE CASACORE - BEGIN
...
-                                                    getHR(ires2).getExprSet(),
-                                                    expr,
-                                                    node.itsIndices1.style()));
+        topStack()->addUpdate (std::shared_ptr<TableParseUpdate>
+                               (new TableParseUpdate(node.itsName,
+                                                     node.itsNameMask,
+                                                     getHR(ires1).getExprSet(),
+                                                     getHR(ires2).getExprSet(),
+                                                     expr,
+                                                     node.itsIndices1.style())));
```
Comment (mnijhuis-tos) [inline tables/TaQL/TaQLNodeHandler.cc:482]: Optional. The other `addUpdate()` calls can also benefit from `std::make_shared`.
```suggestion
        topStack()->addUpdate (std::make_shared<TableParseUpdate>
                                                    (node.itsName,
                                                     node.itsNameMask,
                                                     getHR(ires1).getExprSet(),
                                                     getHR(ires2).getExprSet(),
                                                     expr,
                                                     node.itsIndices1.style()));
```

File: tables/TaQL/TableParseGroupby.cc [hunk 11]
```diff
@@ -225,24 +225,24 @@ namespace casacore { //# NAMESPACE CASACORE - BEGIN
...
     // This function is a special case because it does not need to
     // step though the table. Only its size is of interest. Furthermore,
     // some other columns can also be listed which will be those of the
     // last row.
     // Make a set containing the count(*) aggregate function object.
-    std::vector<CountedPtr<TableExprGroupFuncSet> > funcSets
-      (1, new TableExprGroupFuncSet());
-    CountedPtr<TableExprGroupFuncBase> funcb = itsAggrNodes[0]->makeGroupAggrFunc();
+    std::vector<std::shared_ptr<TableExprGroupFuncSet>> funcSets
+      (1, std::shared_ptr<TableExprGroupFuncSet>(new TableExprGroupFuncSet()));
```
Comment (mnijhuis-tos) [inline tables/TaQL/TableParseGroupby.cc:242]:
```suggestion
      (1, std::make_shared<TableExprGroupFuncSet>());
```

File: tables/TaQL/TableParseGroupby.cc [hunk 12]
```diff
@@ -251,18 +251,18 @@ namespace casacore { //# NAMESPACE CASACORE - BEGIN
       rownrs.reference (Vector<rownr_t>(1, rownrs[rownrs.size()-1]));
     }
     // Save the aggregation results in a result object.
-    return CountedPtr<TableExprGroupResult>(new TableExprGroupResult(funcSets));
+    return std::shared_ptr<TableExprGroupResult>(new TableExprGroupResult(funcSets));
```
Comment (mnijhuis-tos) [inline tables/TaQL/TableParseGroupby.cc:254]:
```suggestion
    return std::make_shared_ptr<TableExprGroupResult>(funcSets);
```

File: tables/TaQL/TableParseProject.cc [hunk 13]
```diff
@@ -869,9 +869,10 @@ namespace casacore { //# NAMESPACE CASACORE - BEGIN
     for (uInt i=0; i<columnExpr_p.size(); i++) {
       if (! columnExpr_p[i].isNull()) {
         if (projectExprSelColumn_p[i] == useSel) {
-          tpq.addUpdate (new TableParseUpdate (columnNames_p[i],
-                                               columnNameMasks_p[i],
-                                               columnExpr_p[i], False));
+          tpq.addUpdate (std::shared_ptr<TableParseUpdate>
+                         (new TableParseUpdate (columnNames_p[i],
+                                                columnNameMasks_p[i],
+                                                columnExpr_p[i], False)));
```
Comment (mnijhuis-tos) [inline tables/TaQL/TableParseProject.cc:875]:
```suggestion
          tpq.addUpdate (std::make_shared<TableParseUpdate>
                                               (columnNames_p[i],
                                                columnNameMasks_p[i],
                                                columnExpr_p[i], False));
```

File: tables/TaQL/test/tExprGroupArray.cc [hunk 14]
```diff
@@ -104,19 +104,18 @@ void checkLazy (const TableExprNode& expr,
     funcid.apply (id);
   }
   funcid.finish();
-  vector<CountedPtr<TableExprGroupFuncSet> > funcSets;
-  funcSets.push_back (new TableExprGroupFuncSet());
-  CountedPtr<TableExprGroupFuncBase> func = aggr.makeGroupAggrFunc();
+  std::vector<std::shared_ptr<TableExprGroupFuncSet>> funcSets;
+  funcSets.push_back (std::shared_ptr<TableExprGroupFuncSet>(new TableExprGroupFuncSet()));
```
Comment (mnijhuis-tos) [inline tables/TaQL/test/tExprGroupArray.cc:108]:
```suggestion
  funcSets.push_back (std::make_shared<TableExprGroupFuncSet>());
```

File: tables/TaQL/test/tExprGroupArray.cc [hunk 15]
```diff
@@ -104,19 +104,18 @@ void checkLazy (const TableExprNode& expr,
...
   if (!allEQ (val.array(), expVal)) {
     foundError = True;
     cout << str << ": found value " << val.array() << "; expected "
          << expVal << endl;
   }
-  vector<CountedPtr<vector<TableExprId> > > ids(1, funcid.getIds());
-  CountedPtr<TableExprGroupResult> groupResult =
-    new TableExprGroupResult(funcSets, ids);
+  std::vector<std::shared_ptr<std::vector<TableExprId>>> ids(1, funcid.getIds());
+  std::shared_ptr<TableExprGroupResult> groupResult(new TableExprGroupResult(funcSets, ids));
```
Comment (mnijhuis-tos) [inline tables/TaQL/test/tExprGroupArray.cc:118]:
```suggestion
  auto groupResult = std::make_shared<TableExprGroupResult>(funcSets, ids);
```

File: tables/Tables/BaseTable.cc [hunk 16]
```diff
@@ -1019,8 +1019,8 @@ BaseTableIterator* BaseTable::makeIterator
 
 const TableDesc& BaseTable::makeEmptyTableDesc() const
 {
-    if (tdescPtr_p.null()) {
-        const_cast<BaseTable*>(this)->tdescPtr_p = new TableDesc();
+    if (!tdescPtr_p) {
+        const_cast<BaseTable*>(this)->tdescPtr_p.reset (new TableDesc());
```
Comment (mnijhuis-tos) [inline tables/Tables/BaseTable.cc:1023]: Please consider removing this `const_cast` in another PR, or, if removing is not possible, document why the `const_cast` is needed in that PR.

File: tables/Tables/ConcatTable.cc [hunk 17]
```diff
@@ -319,14 +319,10 @@ namespace casacore { //# NAMESPACE CASACORE - BEGIN
   void ConcatTable::initialize()
   {
     // Check if all tables have the same description.
-    // Note that we size the table instead of reserve, because push_back
-    // gives the following warning for CountedPtr:
-    //  "dereferencing pointer aonymous  does break strict-aliasing rule"
-    vector<CountedPtr<TableDesc> > actualDesc(tables_p.nelements());;
+    std::vector<std::shared_ptr<TableDesc>> actualDesc(tables_p.nelements());;
```
Comment (mnijhuis-tos) [inline tables/Tables/ConcatTable.cc:322]: Optional: With `std::shared_ptr`, using `push_back()` should be possible again.

File: tables/apps/taql.cc [hunk 18]
```diff
@@ -88,6 +88,21 @@ using namespace std;
 // Define the type for the map of name to (resulttable,command).
 typedef map<String, pair<Table,String> > TableMap;
 
+// Define Deleter objects for a unique_ptr<ostream>
```
Comment (mnijhuis-tos) [inline tables/apps/taql.cc:91]:
```suggestion
// Define Deleter objects for a std::shared_ptr<ostream>
```

**Review by mnijhuis-tos — COMMENTED** [review]
The changes are good, however, I found many spots where the code can be improved using `std::make_shared`.

Lines like `ptr.reset (new Ptr());` could also be converted to `ptr = std::make_shared<Ptr>();` but I did not mark those spots, since `reset()` is also clear and correct.

File: casa/Logging/LogSink.h [hunk 19]
```diff
@@ -302,15 +300,15 @@ class LogSink : public LogSinkInterface
   static void createGlobalSink();
 
   //# Data members.
-  CountedPtr<LogSinkInterface> local_sink_p;
-  static CountedPtr<LsiIntermediate> * global_sink_p;
+  std::shared_ptr<LogSinkInterface> local_sink_p;
+  static std::shared_ptr<LsiIntermediate> * global_sink_p;
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply casa/Logging/LogSink.h:304]: I fully agree. I don't know who wrote this code.

File: casa/Logging/LogSink.h (hunk 1 above)
Comment (gervandiepen) [reply casa/Logging/LogSink.h:311]: I've changed the comment bit. I think the author meant that the order of destruction of global objects is not well defined, so if a LogSink object is used as a global object, the overall global LogSink might be deleted before it.

File: tables/TaQL/ExprNode.cc (hunk 4 above)
Comment (gervandiepen) [reply tables/TaQL/ExprNode.cc:1497]: done

File: tables/TaQL/test/tExprGroupArray.cc (hunk 14 above)
Comment (gervandiepen) [reply tables/TaQL/test/tExprGroupArray.cc:108]: done

File: tables/TaQL/test/tExprGroupArray.cc (hunk 15 above)
Comment (gervandiepen) [reply tables/TaQL/test/tExprGroupArray.cc:118]: done

File: tables/TaQL/TableParseSortKey.cc [hunk 20]
```diff
@@ -76,156 +76,144 @@ namespace casacore { //# NAMESPACE CASACORE - BEGIN
...
-        arrPtr.reset (array);
+        auto array = std::make_shared<Array<Bool>>(node_p.getColumnBool(rownrs));
+        arrPtr = array;
         Bool* data = array->getStorage (deleteIt);
         if (deleteIt) {
           // A copy has been made, so replace the array.
-          arrPtr.reset (new Array<Bool> (array->shape(), data, TAKE_OVER));
+          arrPtr = std::make_shared<Array<Bool>>(array->shape(), data, TAKE_OVER);
         }
         sort.sortKey (data, TpBool, 0, order);
```
Comment (mnijhuis-tos) [inline tables/TaQL/TableParseSortKey.cc:86]: Understanding this code is hard. It looks like the `getStorage` call forces using a contiguous memory chunk, so we can pass the `data` pointer to `sort.sortKey`.

If I'm correct, could you please change this code so it reflects that behavior. For example:
```
arrPtr = std::make_shared<Array<Bool>>(node_p.getColumnBool(rownrs));
if (!arrPtr->contiguousStorage()) arrPtr->unique(); // Make the array contiguous.
sort.sortKey (arrPtr->data(), TpBool, 0, order);
```
Please also update the other `case`s below.

**Review by mnijhuis-tos — APPROVED** [review]
In itself, this PR only has improvements, so I'll approve it.

However, the purpose of the `getStorage()` call is really unclear so please try improving it.

File: tables/TaQL/ExprAggrNodeArray.cc (hunk 3 above)
Comment (gervandiepen) [reply tables/TaQL/ExprAggrNodeArray.cc:74]: Done

File: tables/TaQL/ExprNodeArray.cc (hunk 5 above)
Comment (gervandiepen) [reply tables/TaQL/ExprNodeArray.cc:69]: done

File: tables/TaQL/ExprNodeSet.cc (hunk 7 above)
Comment (gervandiepen) [reply tables/TaQL/ExprNodeSet.cc:220]: done

File: tables/TaQL/ExprNodeSet.cc (hunk 9 above)
Comment (gervandiepen) [reply tables/TaQL/ExprNodeSet.cc:255]: done

File: tables/TaQL/TaQLNodeHandler.cc (hunk 10 above)
Comment (gervandiepen) [reply tables/TaQL/TaQLNodeHandler.cc:482]: Good point. Changed them.

File: tables/TaQL/TableParseGroupby.cc (hunk 12 above)
Comment (gervandiepen) [reply tables/TaQL/TableParseGroupby.cc:254]: done

File: tables/TaQL/TableParseProject.cc (hunk 13 above)
Comment (gervandiepen) [reply tables/TaQL/TableParseProject.cc:875]: done

File: tables/TaQL/TableParseGroupby.cc (hunk 11 above)
Comment (gervandiepen) [reply tables/TaQL/TableParseGroupby.cc:242]: done

File: tables/Tables/ConcatTable.cc (hunk 17 above)
Comment (gervandiepen) [reply tables/Tables/ConcatTable.cc:322]: Knowing the vector size in advance is good; it avoids resizing. vector.reserve could be used with push_back, but sizing the vector with the very cheap shared_ptr constructor seems ok.

File: tables/apps/taql.cc (hunk 18 above)
Comment (gervandiepen) [reply tables/apps/taql.cc]: done

File: tables/TaQL/TableParseSortKey.cc (hunk 20 above)
Comment (gervandiepen) [reply tables/TaQL/TableParseSortKey.cc:86]: Avoid getStorage by testing if the array is contiguous.

File: tables/TaQL/ExprNodeSet.cc (hunk 8 above)
Comment (gervandiepen) [reply tables/TaQL/ExprNodeSet.cc:247]: done

File: tables/TaQL/ExprNodeRep.h (hunk 6 above)
Comment (gervandiepen) [reply tables/TaQL/ExprNodeRep.h:55]: It is used in many places. I think it makes the code easier to read because it (IMHO) makes it easier to grasp a statement.

File: tables/Tables/BaseTable.cc (hunk 16 above)
Comment (gervandiepen) [reply tables/Tables/BaseTable.cc]: I agree const_cast is not nice and should be removed. However, it requires quite some functions to be made non-const, which is too much for now.
I plan to go through all const_cast in casacore and see if they can be changed.

**Comment by gervandiepen** [comment]
Thanks for your thorough review and hints.
